src/libcollections/string.rs

```diff
@@ -1899,3 +1956,74 @@ impl<'a> DoubleEndedIterator for Drain<'a> {
         self.iter.next_back()
     }
 }
+
+/// A splicing iterator for `String`.
+///
+/// This struct is created by the [`splice()`] method on [`String`]. See its
+/// documentation for more.
+///
+/// [`splice()`]: struct.String.html#method.splice
+/// [`String`]: struct.String.html
+#[unstable(feature = "splice", reason = "recently added", issue = "32310")]
+pub struct Splice<'a, 'b> {
+    /// Will be used as &'a mut String in the destructor
+    string: *mut String,
+    /// Start of part to remove
+    start: usize,
+    /// End of part to remove
+    end: usize,
+    /// Current remaining range to remove
+    iter: Chars<'a>,
+    replace_with: &'b str,
+}
+
+#[unstable(feature = "splice", reason = "recently added", issue = "32310")]
+unsafe impl<'a, 'b> Sync for Splice<'a, 'b> {}
+#[unstable(feature = "splice", reason = "recently added", issue = "32310")]
+unsafe impl<'a, 'b> Send for Splice<'a, 'b> {}
+
+#[unstable(feature = "splice", reason = "recently added", issue = "32310")]
+impl<'a, 'b> Drop for Splice<'a, 'b> {
+    fn drop(&mut self) {
+        unsafe {
```

Comment:
Another detail is that `String::splice` returns an `Iterator<Item=char>`, which is not easy to implement on top of `Iterator<Item=u8>` returned by splice on the underlying bytes vector.

But that iterator is not as useful as on `Vec<T>` where `T` might not be `Copy`. You can always access `string[range]` before calling `string.splice(range, replacement)`. (And maybe even skip the UTF-8 round-trip if you end up using `.to_owned()` on that slice instead of `.collect()` on a `char` iterator.) So I wouldn’t mind having `String::splice` return nothing, in which case it is trivial to implement with `Vec::splice`. This even is how I wrote the RFC since I’d assumed (and didn’t check) that `String::drain` also returns nothing.